sdk/metrics/src/main/java/io/opentelemetry/sdk/metrics/export/PeriodicMetricReader.java

```diff
@@ -0,0 +1,191 @@
+import io.opentelemetry.sdk.internal.DaemonThreadFactory;
+import java.time.Duration;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+/**
+ * Wraps a {@link MetricExporter} and automatically reads and exports the metrics every export
+ * interval. Metrics may also be dropped when it becomes time to export again, and there is an
+ * export in progress.
+ */
+public class PeriodicMetricReader implements MetricReader {
+  private static final Logger logger = Logger.getLogger(PeriodicMetricReader.class.getName());
+
+  private final MetricProducer producer;
+  private final MetricExporter exporter;
+  private final ScheduledExecutorService scheduler;
+
+  private final Scheduled scheduled;
+  private volatile ScheduledFuture<?> scheduledFuture;
+  private final Object lock = new Object();
+
+  private PeriodicMetricReader(
+      MetricProducer producer, MetricExporter exporter, ScheduledExecutorService scheduler) {
+    this.producer = producer;
+    this.exporter = exporter;
+    this.scheduler = scheduler;
+    this.scheduled = new Scheduled();
+  }
+
+  @Override
+  public CompletableResultCode flush() {
+    return scheduled.doRun();
+  }
+
+  @Override
+  public CompletableResultCode shutdown() {
+    final CompletableResultCode result = new CompletableResultCode();
+    if (scheduledFuture != null) {
+      scheduledFuture.cancel(false);
+    }
+    scheduler.shutdown();
+    try {
+      scheduler.awaitTermination(5, TimeUnit.SECONDS);
+      final CompletableResultCode flushResult = scheduled.doRun();
+      flushResult.join(5, TimeUnit.SECONDS);
+    } catch (InterruptedException e) {
+      // force a shutdown if the export hasn't finished.
+      scheduler.shutdownNow();
+      // reset the interrupted status
+      Thread.currentThread().interrupt();
+    } finally {
+      final CompletableResultCode shutdownResult = scheduled.shutdown();
+      shutdownResult.whenComplete(
+          () -> {
+            if (!shutdownResult.isSuccess()) {
+              result.fail();
+            } else {
+              result.succeed();
+            }
+          });
+    }
+    return result;
+  }
+
+  private void start(Duration duration) {
+    // Autoconfigure sends us null or 0 durations and expects us to just not start.
+    if (duration == null || duration.isZero()) {
+      return;
+    }
+
+    synchronized (lock) {
+      if (scheduledFuture != null) {
+        return;
+      }
+      scheduledFuture =
+          scheduler.scheduleAtFixedRate(
+              scheduled, duration.toMillis(), duration.toMillis(), TimeUnit.MILLISECONDS);
+    }
+  }
+
+  /** Factory for {@link PeriodicMetricReader}. */
+  public static class Factory implements MetricReader.Factory {
```

Comment:
pull this out to a top level for consistency with the rest of the SDK.

Reply:
fixed